src/vm/amd64/jithelpers_fast.S

```diff
@@ -329,3 +329,152 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
+
+        // we only want the lower 32-bits of rsi, it might be dirty
+        or      esi, esi
+
+        // check that index is in bounds
+        cmp     esi, dword ptr [rdi + OFFSETOF__PtrArray__m_NumComponents] // 8h -> array size offset
+        jae     6f // ThrowIndexOutOfRangeException
+
+        // r10 = Array MT
+        mov     r10, [rdi]
+
+        // if we're assigning a null object* then we don't need a write barrier
+        test    rdx, rdx
+        jz      3f // AssigningNull
+
+#ifdef CHECK_APP_DOMAIN_LEAKS     
+        // get Array TypeHandle
+        mov     rcx, [r10 + OFFSETOF__MethodTable__m_ElementType]   // 10h -> typehandle offset, 
+        // check for non-MT
+        test    rcx, 2
+        jnz     1f // NoCheck
+
+        // Check VMflags of element type
+        mov     rcx, [rcx + OFFSETOF__MethodTable__m_pEEClass]
+        mov     ecx, dword ptr [rcx + OFFSETOF__EEClass__m_wAuxFlags]
+        test    ecx, EEClassFlags
+        jnz     C_FUNC(ArrayStoreCheck_Helper)
+
+    1:// NoCheck:
+#endif
+
+        mov     rcx, [r10 + OFFSETOF__MethodTable__m_ElementType]   // 10h -> typehandle offset
+
+        // check for exact match
+        cmp     rcx, [rdx]
+        jne     4f // NotExactMatch
+
+    2:// DoWrite:
+        lea     rdi, [rdi + 8*rsi]
+        add     rdi, OFFSETOF__PtrArray__m_Array
+        mov     rsi, rdx
+
+        // JIT_WriteBarrier(Object** dst, Object* src)
+        jmp     C_FUNC(JIT_WriteBarrier)
+
+    3:// AssigningNull:
+        // write barrier is not needed for assignment of NULL references
+        mov     [rdi + 8*rsi + OFFSETOF__PtrArray__m_Array], rdx
+        ret
+            
+    4:// NotExactMatch:
+        PREPARE_EXTERNAL_VAR g_pObjectClass, r11
+        cmp     rcx, [r11]
+        je      2b // DoWrite
+
+        jmp     C_FUNC(JIT_Stelem_Ref__ObjIsInstanceOfNoGC_Helper)
+                           
+    5:// ThrowNullReferenceException:
+        mov     rdi, CORINFO_NullReferenceException_ASM 
+        jmp     C_FUNC(JIT_InternalThrow)
+        
+    6:// ThrowIndexOutOfRangeException:
+        mov     rdi, CORINFO_IndexOutOfRangeException_ASM
+        jmp     C_FUNC(JIT_InternalThrow)        
+LEAF_END JIT_Stelem_Ref, _TEXT
+
+LEAF_ENTRY JIT_Stelem_Ref__ObjIsInstanceOfNoGC_Helper, _TEXT
+        push_nonvol_reg rbp
+        mov     rbp, rsp
+        set_cfa_register rbp, 16
+        
+        sub     rsp, 0x20
+        mov     [rbp - 0x08], rdi
+        mov     [rbp - 0x10], rsi
+        mov     [rbp - 0x18], rdx
+
+        // need to get TypeHandle before setting rcx to be the Obj* because that trashes the PtrArray*
+        mov     rsi, rcx
+        mov     rdi, rdx
+
+        // TypeHandle::CastResult ObjIsInstanceOfNoGC(Object *pElement, TypeHandle toTypeHnd)
+        call    C_FUNC(ObjIsInstanceOfNoGC)
+
+        mov     rdi, [rbp - 0x08]
+        mov     rsi, [rbp - 0x10]
+        mov     rdx, [rbp - 0x18]
+        
+        mov     rsp, rbp
```

Comment:
It would be nice to replace this whole block (4 lines) with macro, named e.g. "RESET_FRAME" or something  similar and use it also at the other place before.